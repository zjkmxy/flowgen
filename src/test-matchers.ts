import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
