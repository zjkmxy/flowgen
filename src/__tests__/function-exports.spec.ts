import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle exported es module functions", async () => {
  const ts = `
class Action {}
class RouterState {}
class Store<T> {}
class SyncHistoryWithStoreOptions {}
class HistoryUnsubscribe {}

export function routerReducer(state?: RouterState, action?: Action): RouterState;
export function syncHistoryWithStore(history: History, store: Store<any>, options?: SyncHistoryWithStoreOptions): History & HistoryUnsubscribe;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle toString function overload", async () => {
  const ts = `export function toString(): void;
export function toString(e: number): void;
export function toString(b: string): void;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle default exported es module functions", async () => {
  const ts = `
class Action {}
class RouterState {}

export default function routerReducer(state?: RouterState, action?: Action): RouterState;`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle function overload es module functions", async () => {
  const ts = `
class Action {}
class RouterState {}

export function routerReducer(state?: RouterState, action?: Action): RouterState;
export function routerReducer(state?: RouterState): RouterState;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should remove this annotation from functions", async () => {
  const ts =
    "function addClickListener(onclick: (this: void, e: Event) => void): void;";
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should remove default parameters from functions", async () => {
  const ts =
    "function addClickListener<T = Error>(onclick: (e: Event) => void): T;";
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should not break with Promise return types - issue 156", async () => {
  const ts = "export declare const fn: () => Promise<void>;";
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
