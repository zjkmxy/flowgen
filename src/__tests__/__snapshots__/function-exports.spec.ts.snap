// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle default exported es module functions 1`] = `
"declare class Action {}
declare class RouterState {}
declare export default function routerReducer(
  state?: RouterState,
  action?: Action,
): RouterState;
"
`;

exports[`should handle exported es module functions 1`] = `
"declare class Action {}
declare class RouterState {}
declare class Store<T> {}
declare class SyncHistoryWithStoreOptions {}
declare class HistoryUnsubscribe {}
declare export function routerReducer(
  state?: RouterState,
  action?: Action,
): RouterState;
declare export function syncHistoryWithStore(
  history: History,
  store: Store<any>,
  options?: SyncHistoryWithStoreOptions,
): History & HistoryUnsubscribe;
"
`;

exports[`should handle function overload es module functions 1`] = `
"declare class Action {}
declare class RouterState {}
declare export function routerReducer(
  state?: RouterState,
  action?: Action,
): RouterState;
declare export function routerReducer(state?: RouterState): RouterState;
"
`;

exports[`should handle toString function overload 1`] = `
"declare export function toString(): void;
declare export function toString(e: number): void;
declare export function toString(b: string): void;
"
`;

exports[`should not break with Promise return types - issue 156 1`] = `
"declare export var fn: () => Promise<void>;
"
`;

exports[`should remove default parameters from functions 1`] = `
"declare function addClickListener<T>(onclick: (e: Event) => void): T;
"
`;

exports[`should remove this annotation from functions 1`] = `
"declare function addClickListener(onclick: (e: Event) => void): void;
"
`;
